### PR TITLE
Cleanup, refactor and cover with tests section-config module

### PR DESCRIPTION
**Pull request by krzksz**
### Description (*)
I decided to create this PR as a follow up to my [previous attempt](https://github.com/magento/magento2/pull/24847) at fixing errors that were caused by 3rd party modules doing POST requests or requesting customer data directly before the module was initialized by `data-mage-init`.

I also added unit tests for all the public methods and to make sure no errors are thrown if some code incorrectly calls the module before is has a chance to initialize. 

### Fixed Issues (if relevant)

1. magento/magento2#17125: x-magento-init initialisation not bound to happen in the right order.

### Manual testing scenarios (*)

1. Reload the shop.
2. Trigger customer data or sections data module at the very beginning for example by running the following code:
```javascript
require(['Magento_Customer/js/section-config'], function(sectionsConfig) {
    console.log(sectionsConfig.getAffectedSections('test'));
});
```
3. `TypeError: sections is undefined` is being thrown without this changes.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
Hi @krzksz,
Please check functional tests failures, looks like your changes break some functionality.

**Comment by krzksz** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @krzksz. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @krzksz, here is your [Magento instance](https://i-25764-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-25764-2-3-develop.instances.magento-community.engineering/admin](https://i-25764-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by krzksz** [comment]
@magento run all tests

**Review by ptylek — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ptylek, thank you for the review.
[ENGCOM-6666](https://jira.corp.magento.com/browse/ENGCOM-6666) has been created to process this Pull Request
:eight_spoked_asterisk: @ptylek, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed 

**Before:**
before

**After:**
after

File: dev/tests/js/jasmine/tests/app/code/Magento/Customer/frontend/js/section-config.test.js [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+                            'all'
+                        ]
+                    },
+                    baseUrls: [
+                        'http://localhost.com/'
+                    ]
+                });
+
+                expect(obj.getAffectedSections('http://localhost.com/path')).toEqual(['section', 'all']);
+            });
```
Comment (omiroshnichenko) [inline dev/tests/js/jasmine/tests/app/code/Magento/Customer/frontend/js/section-config.test.js]: This test is the same as https://github.com/magento/magento2/pull/25764/files#diff-b2f763698c12c5411f66d86c6390b5adR99-R114

**Review by omiroshnichenko — CHANGES_REQUESTED** [review]
1 tiny comment, but everything else looks good. Thank you for cleanup and especially for covering it with the unit test.

File: dev/tests/js/jasmine/tests/app/code/Magento/Customer/frontend/js/section-config.test.js (hunk 1 above)
Comment (krzksz) [reply dev/tests/js/jasmine/tests/app/code/Magento/Customer/frontend/js/section-config.test.js]: True, will fix ASAP.

Comment (krzksz) [reply dev/tests/js/jasmine/tests/app/code/Magento/Customer/frontend/js/section-config.test.js]: Fixed.

**Review by omiroshnichenko — APPROVED** [review]

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-6666](https://jira.corp.magento.com/browse/ENGCOM-6666) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Comment by engcom-Kilo** [comment]
Functional tests CE - not successful on remote test server,
but not depends by changes in Pull Request.
Remote Test server - with some fails with Create test data.

**Pull request merged**